### PR TITLE
Add a config option for validating 'next_link' parameters against a domain whitelist

### PR DESCRIPTION
**Pull request by anoadragon453**
This is a config option ported over from DINUM's Sydent: https://github.com/matrix-org/sydent/pull/285

They've switched to validating 3PIDs via Synapse rather than Sydent, and would like to retain this functionality.

This original purpose for this change is phishing prevention. This solution could also potentially be replaced by a similar one to https://github.com/matrix-org/synapse/pull/8004, but across all `*/submit_token` endpoint.

This option may still be useful to enterprise even with that safeguard in place though, if they want to be absolutely sure that their employees don't follow links to other domains.

### CODE REVIEW

File: synapse/rest/client/v2_alpha/account.py [hunk 1]
```diff
@@ -603,15 +617,10 @@ async def on_GET(self, request):
 
             # Perform a 302 redirect if next_link is set
             if next_link:
-                if next_link.startswith("file:///"):
-                    logger.warning(
-                        "Not redirecting to next_link as it is a local file: address"
-                    )
```
Comment (anoadragon453) [inline synapse/rest/client/v2_alpha/account.py:609]: This check now happens implicitly in `assert_valid_next_link`, as only `http` and `https` `next_link` values are allowed.

Additionally, `next_link` can only be specified during the call to `/requestToken`. Having this check in `/submit_token`, blocking the user after already approving their `/requestToken` call, is not great UX.

File: synapse/rest/client/v2_alpha/account.py [hunk 2]
```diff
@@ -875,6 +884,45 @@ async def on_POST(self, request):
...
+        SynapseError: If the next_link is invalid
+    """
+    valid = True
+
+    # Parse the contents of the URL
+    next_link_parsed = urlparse(next_link)
+
+    # Scheme must be http(s)
+    if next_link_parsed.scheme not in ["http", "https"]:
+        valid = False
```
Comment (anoadragon453) [inline synapse/rest/client/v2_alpha/account.py:908]: I wonder whether this is too restrictive. Someone may wish to redirect to some otherprotocol (`ipfs://`, `dat://`, etc.). Perhaps this should still just disallow `file:///` again.

File: synapse/config/server.py [hunk 3]
```diff
@@ -542,6 +542,14 @@ class LimitRemoteRoomsConfig:
             users_new_default_push_rules
         )  # type: set
 
+        # Whitelist of domain names that given next_link parameters must have
+        next_link_domain_whitelist = config.get("next_link_domain_whitelist")
+        if not isinstance(next_link_domain_whitelist, list):
```
Comment (babolivier) [inline synapse/config/server.py:547]: Surely this will fail if the setting isn't provided (which isn't what the doc in the config says)? Given `d.get` returns `None` by default if the key wasn't found.

File: synapse/config/server.py [hunk 4]
```diff
@@ -1014,6 +1022,24 @@ def generate_config_section(
...
+        # must match.
+        #
+        # This parameter is optionally provided by clients while requesting
+        # validation of an email or phone number, and maps to a link that
+        # users will be automatically redirected to after validation
+        # succeeds. Clients can make use this parameter to aid the validation
+        # process.
+        #
+        # The whitelist is applied whether the homeserver or an
+        # account_threepid_delegate is handling validation.
```
Comment (babolivier) [inline synapse/config/server.py:1036]:
```suggestion
        # identity server is handling validation.
```

File: synapse/rest/client/v2_alpha/account.py (hunk 2 above)
Comment (babolivier) [reply synapse/rest/client/v2_alpha/account.py:908]: I think it's better to just disallow `file:///` but ymmv.

File: synapse/config/server.py (hunk 4 above)
Comment (anoadragon453) [reply synapse/config/server.py:1036]: Heh, I actually wrote this so that when we (hopefully) remove the `account_threepid_delegate` option, I remember to change this docstring as well.

You can *technically* handle verification without a full identity server, but just something that handles the verification endpoints. I've been careful to avoid writing "identity server" in other places because of this.

File: synapse/rest/client/v2_alpha/account.py (hunk 2 above)
Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/account.py:908]: Yeah, I don't want to break someone's Riot hosted inside a firefox extension (`moz-extension://` or something). I'll change it back.

File: synapse/config/server.py (hunk 4 above)
Comment (babolivier) [reply synapse/config/server.py:1041]: According to its documentation, `account_threepid_delegate` only works if the address matches a trusted identity server. Also, one could argue that something that implements the verification endpoint is _technically_ an identity server, it just doesn't implement the IS spec fully, but can still be used to manage and verify identity to some extent.
Though more to the point, my concern is that "an account_threepid_delegate" doesn't really mean anything and is pretty confusing imho.

File: tests/rest/client/v2_alpha/test_account.py [hunk 5]
```diff
@@ -668,16 +669,79 @@ def test_no_valid_token(self):
...
+        )
+
+    def _request_token(
+        self,
+        email: str,
+        client_secret: str,
+        next_link: Optional[str] = None,
+        expect_code: int = 200,
+    ) -> str:
+        """Request validating and adding an email to a user account
```
Comment (babolivier) [inline tests/rest/client/v2_alpha/test_account.py:723]: I find this a bit hard to parse, maybe

```suggestion
        """Request a validation token to add an email address to a user's account
```

would work better?

File: tests/rest/client/v2_alpha/test_account.py [hunk 6]
```diff
@@ -668,16 +669,79 @@ def test_no_valid_token(self):
...
+        )
+
+        self._request_token(
+            "something@example.com",
+            "some_secret",
+            next_link="https://bad.example.org/some/bad/page",
+            expect_code=400,
+        )
+
+    @override_config({"next_link_domain_whitelist": []})
```
Comment (babolivier) [inline tests/rest/client/v2_alpha/test_account.py:706]: It's possibly worth (and trivial) to add a test for when this is `None` (e.g. no whitelist)?

**Review by babolivier — CHANGES_REQUESTED** [review]
Couple of nitpicks plus https://github.com/matrix-org/synapse/pull/8275#discussion_r484801289 which didn't make it into the review for some reason.

File: synapse/config/server.py (hunk 4 above)
Comment (anoadragon453) [reply synapse/config/server.py:1041]: Wow, yeah my bad, the docs for `account_threepid_delegates` mention identity servers all over the place :| Also that bit is a bit confusing, as we have another option for trusted identity servers, which isn't relevant here.

And yeah, good point. They may be very small processes, but they're still using the IS API.

**Review by babolivier — APPROVED** [review]
lgtm, modulo CI
